Refactor storyboard commands structure and add framework-based transform loop support

frenzibyte opened the pull request:
- Closes https://github.com/ppy/osu/issues/12401
- Supersedes & closes https://github.com/ppy/osu/pull/27482 

The need for a refactor sparked while I was reviewing https://github.com/ppy/osu/pull/27482, a PR that intended to replace the ad-hoc transform loop logic with an actual framework-based looping transform sequence. 

The PR has shown that the current structure of the storyboard commands flow was not flexible enough to have looping easily supported, for multiple reasons:
 1. A transform sequence is required for `GeneratedCommand<T>` to append a loop operation on the transforms, but it's not possible to return a transform sequence from the transform delegates, unless another generic parameter is added to the delegate, making code extremely awful.
     - attempted somewhat in previous PR: (https://github.com/ppy/osu/pull/27482/commits/8b03acd27bdbdd6a997dbd70a00a7b4568e6b8bc)
 2. Moving the transform application logic to the `GeneratedCommand<T>` will avoid the necessity of adding generic parameters, but at the cost of using `TransformTo` and tracking both the property name and property grouping (technical detail in `Transform`s).
     - attempted in previous PR: (https://github.com/ppy/osu/pull/27482/commits/c05007804fb715b5d542535bcb224d5bd67b8b58 + https://github.com/ppy/osu/pull/27482/commits/7193ec66a4dee4abe516bd8984d3ed50559df35d)

And generally, I did not like that everything was being served to `StoryboardSprite` as flat data, and `StoryboardSprite` taking on the role of doing literally everything by itself. If I understand correctly, that shows a lack of encapsulation and polymorphism, destroying code readability as well as quality/flexibility.

My refactor intends to address all of the problems above by removing the command generation logic in `ApplyTransforms` and use polymorphism for transform application (instead of handling that case-by-case in `StoryboardSprite.ApplyTransforms` and the helper methods surrounding it). The refactor also simplifies most of the structure and merges `CommandTimeline` and `CommandTimelineGroup` together (removing the iteration-based properties in `CommandTimelineGroup` in the process).

In the process, I've already supported framework-based transform loops in this PR. I can split it to a second PR, but I'm not sure there's much reason to. If it helps reviewing and ensuring stability, I can split it out.

Feedback on the efficiency of the new structure with respect to performance would be appreciated, and/or an idea on how to make benchmarks for this. I'm not quite familiar with the performance side of things, but I hope the new structure isn't overly inefficient for some reason.

I've also done a general pass on the naming of all relevant classes to read better for me. Terms like `Timeline` combined with `Group` makes me mistake the classes for something entirely different. Open for feedback.

EVAST9919 committed:
Use more direct way to apply transforms

EVAST9919 committed:
Make use of framework's transform loops

EVAST9919 committed:
Remove unused transform extensions

EVAST9919 committed:
Fix broken tests

EVAST9919 committed:
Implement StoryboardElementWithDuration

EVAST9919 committed:
Remove blank lines

EVAST9919 committed:
Further simplify transform application

frenzibyte committed:
Apply major refactor to the storyboard commands flow structrure

frenzibyte committed:
Calculate loop delays at point of transform application

frenzibyte committed:
Fix group start/end time not calculating correctly

frenzibyte committed:
Fix `HasCommands` property not set at all

frenzibyte committed:
Fix incorrect start time calculations

`LoopStartTime` is now baked into each `IStoryboardCommand`.

frenzibyte committed:
Fix comparison interface not implemented on storyboard command classes

frenzibyte committed:
Support applying initial values of storyboard commands

frenzibyte committed:
Add generics to `ApplyInitialValue`/`ApplyTransforms` for ability to return custom transform sequences

*sigh*

frenzibyte committed:
Fix refactor error on `EndTimeForDisplay`

Not sure when this happened >.>

frenzibyte committed:
Merge branch 'master' into storyboard-loops-v2

frenzibyte committed:
Fix merge conflicts

frenzibyte committed:
Reorder command properties to match general format

I had them shuffled around in the middle of the refactor.

frenzibyte committed:
Fix parameter commands applying initial value before start time

frenzibyte committed:
Expose storyboard command lists as read-only and remove unnecessary memory footprint

Mutation should be done only with the methods exposed by `StoryboardCommandGroup`.

frenzibyte committed:
Remove leftover debugging code

frenzibyte committed:
Remove unused using directive

frenzibyte committed:
Remove extra end line

frenzibyte committed:
Add basic test scene for asserting storyboard commands behaviour

Pending actual test coverage.

frenzibyte committed:
Adjust time values

frenzibyte committed:
Add test case for running with high number of loops

frenzibyte committed:
Add enough test coverage

frenzibyte commented:
As an update, this PR is ready for review. Would appreciate testing with any storyboards that may have potential edge cases. I've already tested thoroughly with [world.execute(me)](https://osu.ppy.sh/beatmapsets/470977#osu/1012836) and [Ayane - Nageki no Mori](https://osu.ppy.sh/beatmapsets/1108479#osu/2316652), and while both of them look to work correctly according to my eyesight, a second pair would be helpful.

bdach commented on osu.Game/Storyboards/Commands/IStoryboardCommand.cs:
```diff
@@ -0,0 +1,46 @@
+        /// Sets the value of the corresponding property in <see cref="Drawable"/> to the start value of this command.
+        /// </summary>
+        /// <remarks>
+        /// Parameter commands (e.g. <see cref="StoryboardFlipHCommand"/> / <see cref="StoryboardFlipVCommand"/> / <see cref="StoryboardBlendingParametersCommand"/>) only apply the start value if they have zero duration, i.e. take "permanent" effect regardless of time.
```
This is gonna require elaboration because I'm not sure what that means. Is this some weird stable quirk? As I understand those commands they flip the drawable / change blending of the drawable between `StartTime` and `EndTime`, so shouldn't a zero-duration "parameter command" do... nothing ever?

Walavouchey replied:
no, any command with zero duration takes permanent effect

1. the end value is used (not the start value) after the time of the command
2. the start value is used before the time of the command (if there are no commands of the same type before it)

bdach commented on osu.Game/Storyboards/Commands/StoryboardCommand.cs:
```diff
@@ -0,0 +1,56 @@
+
+        public abstract string PropertyName { get; }
+
+        public abstract void ApplyInitialValue<TDrawable>(TDrawable d)
+            where TDrawable : Drawable, IFlippable, IVectorScalable;
+
+        public abstract TransformSequence<TDrawable> ApplyTransforms<TDrawable>(TDrawable d)
+            where TDrawable : Drawable, IFlippable, IVectorScalable;
```
Rather than have generics here I'd prefer what https://github.com/ppy/osu/pull/27482 did and have `IDrawableStoryboardElement : IFlippable, IVectorScalable` if possible. I just don't really see that much point here.

(I did initially say in that pull that that felt arbitrary but I guess this explains it better than that diff ever did.)

frenzibyte replied:
It's not possible to use any interface as a generic type argument for `TransformSequence` because it's constrained to `Drawable` classes, that's why I have the generic parameter here constrained to `Drawable`s.

We can change the `TransformSequence` constraint to `IDrawable`, but I'm not sure if the effort is worth it, so I bailed out and added a *pretty local* generic here.

bdach commented on osu.Game/Storyboards/Commands/StoryboardCommandGroup.cs:
```diff
@@ -0,0 +1,130 @@
+        public double EndTime { get; private set; } = double.MinValue;
+
+        [JsonIgnore]
+        public double Duration => EndTime - StartTime;
```
I'm not sure what to think of the fact that by default a `StoryboardCommandGroup` appears to have a `Duration` of negative infinity.

frenzibyte replied:
yeah this has been a thing in master as well, I can change to null but then I have to handle null `StartTime`/`EndTime` all over the storyboard sprite class sooo.

bdach commented on osu.Game/Storyboards/Commands/StoryboardLoopingGroup.cs:
```diff
@@ -0,0 +1,66 @@
+        /// <param name="repeatCount">The number of times the loop should repeat. Should be greater than zero. Zero means a single playback.</param>
+        public StoryboardLoopingGroup(double startTime, int repeatCount)
+        {
+            if (repeatCount < 0) throw new ArgumentException("Repeat count must be zero or above.", nameof(repeatCount));
```
```suggestion
            ArgumentOutOfRangeException.ThrowIfNegative(repeatCount);
```

bdach commented on osu.Game/Storyboards/Commands/StoryboardTriggerGroup.cs:
```diff
@@ -1,16 +1,16 @@
+namespace osu.Game.Storyboards.Commands
 {
-    public class CommandTrigger : CommandTimelineGroup
+    public class StoryboardTriggerGroup : StoryboardCommandGroup
```
What's with this rename? A "trigger group" does not seem like it could ever be the same thing as a "command trigger"...?

Walavouchey replied:
a "trigger group" (people just call the triggers) is a group of commands that all activate when the trigger condition is met. the commands within a trigger are indented in the .osu/.osb files. it makes conceptual sense if anything

frenzibyte replied:
As per the above comment, the term "trigger group" is better at describing how this class functions than "command trigger" and remains consistent with loop groups name-wise (`StoryboardLoopingGroup`).

bdach commented on osu.Game/Storyboards/StoryboardSprite.cs:
```diff
@@ -118,94 +115,43 @@ public StoryboardSprite(string path, Anchor origin, Vector2 initialPosition)
-        }
-
-        private void generateCommands<T>(List<IGeneratedCommand> resultList, IEnumerable<CommandTimeline<T>.TypedCommand> commands,
-                                         DrawablePropertyInitializer<T> initializeProperty, DrawableTransformer<T> transform, bool alwaysInitialize = true)
-        {
-            bool initialized = false;
-
-            foreach (var command in commands)
+            foreach (var command in commands.OrderBy(c => c.StartTime))
```
This looks... expensive...

- `AllCommands`, which is a `SelectManyIterator`
- which is then linq-concatenated with another `SelectManyIterator`
- and then finally linq-sorted

Was linq overhead measured here? It looks like it would be massive. I guess this is only ran during load which makes it less of a killer, but surely this can be improved by materialising somewhere?

frenzibyte replied:
I have been going quite easy on the code because it's a once-off, yeah. I will revise it.

bdach reviewed (commented):
basic structure pass. a lot more palatable structurally than https://github.com/ppy/osu/pull/27482 but i've got the same issue here of not actually being able to see where the fix is super well

frenzibyte commented on osu.Game/Storyboards/Commands/StoryboardLoopingGroup.cs:
```diff
@@ -0,0 +1,66 @@
+
+            public override void ApplyInitialValue<TDrawable>(TDrawable d) => command.ApplyInitialValue(d);
+
+            public override TransformSequence<TDrawable> ApplyTransforms<TDrawable>(TDrawable d)
+            {
+                if (loopingGroup.TotalIterations == 0)
+                    return command.ApplyTransforms(d);
+
+                double loopingGroupDuration = loopingGroup.Duration;
+                return command.ApplyTransforms(d).Loop(loopingGroupDuration - Duration, loopingGroup.TotalIterations);
+            }
```
@bdach if I understand what you are asking, the fix lies here specifically. normally, with my refactor but without the fix applied, this code would be a for-loop manually applying the transform until `TotalIteration` times (causing memory allocation issues etc.).

bdach commented:
Merge conflicts are probably going to be "fun" to resolve here.

Coppertine commented:
> As an update, this PR is ready for review. Would appreciate testing with any storyboards that may have potential edge cases. I've already tested thoroughly with [world.execute(me)](https://osu.ppy.sh/beatmapsets/470977#osu/1012836) and [Ayane - Nageki no Mori](https://osu.ppy.sh/beatmapsets/1108479#osu/2316652), and while both of them look to work correctly according to my eyesight, a second pair would be helpful.

Large post incoming..
Transformations look pretty good on my end for [uncanny long arms](https://osu.ppy.sh/beatmapsets/2094154#osu/4390731) (one of the `vig.png` sprites are small by like 1px on the sides of the display, but it's more so a self made scaling issue (`0.444` instead of `0.444444444`)), [Reality Distortion](https://osu.ppy.sh/beatmapsets/582089#osu/1480798) and the Liquicity Yearmix storyboard (albiet to some lag, but is expected, and you can not change the current time on the timeline in auto due to how the stars are setup so i might refactor the storyboard). as well as some other storyboards I have tried out with.
However i would like to note that [NOMA - Louder Machine](https://osu.ppy.sh/beatmapsets/483606#osu/1042243) (due to the second half) likes to run at around 30fps (on a 144hz monitor set to 8x refresh), possibly due to the ~170k commands / ~10k sprites in that section.

Another thing I noticed was that easings on opacity changes on sprites don't appear during gameplay.
Here's a sprite code of one of my storyboards ([Astsumei - Tenshi no Kikyou (feat. L4hee)](https://osu.ppy.sh/beatmapsets/2036657#osu/4247350)) for example. (it applies to other storyboards, but this one's the easiest to examine)
```
Sprite,Foreground,Centre,"sb/p.png",320,240
 V,0,87333,,853,480
 F,19,87333,88333,1,0
 F,16,89999,90415,1,0
 F,16,90499,90999,1,0
 F,16,90999,91665,1,0
 F,19,103332,103532,1,0
 F,19,108665,109165,1,0
 F,15,71333,72833,1,0
 F,18,92665,93499,1,0
 F,19,183332,186332,1,0
 F,19,195665,197665,1,0
 F,19,199998,201998,1,0
 F,19,203332,203593,1,0
 F,19,203679,203979,1,0
 F,19,204080,206080,1,0
 F,19,121999,123999,1,0
 F,19,113999,115999,1,0
 F,19,124665,134665,1,0
 F,19,177998,187998,1,0
 F,0,295430,296430,1,0
 F,19,300763,301763,1,0
 F,19,303430,304430,1,0
 F,19,306096,309096,1,0
 F,19,348763,351763,1,0
 F,19,393130,403130,1,0
 ```

frenzibyte committed:
Merge branch 'master' into storyboard-loops-v2

frenzibyte committed:
Resolve post-merge-conflict issues

frenzibyte committed:
Make video sprites flippable and vector-scalable to fix type constraint errors

Stable supports these on videos already from a quick read on code (since videos are generally a subclass of sprites).

frenzibyte commented:
@Coppertine I cannot find anything odd with the beatmap you attached, I've checked on one of the sprites of the storyboard that changes opacity, and the easing is populated correctly:

![CleanShot 2024-05-02 at 00 25 53](https://github.com/ppy/osu/assets/22781491/1138003b-386d-4971-9103-395ff18b70fd)

If the issue happens on master as well, it should be taken to an issue thread to be tackled separately.

frenzibyte committed:
Use throw helper methods

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

frenzibyte commented on osu.Game/Storyboards/StoryboardSprite.cs:
```diff
@@ -61,8 +62,8 @@ public double EarliestTransformTime
-                foreach (var l in loops)
+                double earliestStartTime = Commands.StartTime;
+                foreach (var l in loopingGroups)
                     earliestStartTime = Math.Min(earliestStartTime, l.StartTime);
```
Previously `l.StartTime` referred to the start time of the loop command offset from the start time of the loop group, so the loop group start time was supposed to be added to `l.StartTime` in here for `earliestStartTime` to correctly return the start time of the command...

The behaviour is now different in this PR because `l.StartTime` correctly refers to the absolute start time of the command, making this code more correct.

Coppertine commented:
Apologies, opacity changing sprites was fixed in a later commit. Edited to reflect current pr state.

bdach commented:
I dunno what to do with this pull. The size of the diff and the complexities involved make it literally unreviewable as far as my powers go. About all I have to go on is that it does appear to fix the game gobbling up memory on beatmaps with those stupid loop counts and @Coppertine's empirical testing. But having the game not die on OOM does seem like a major positive so I'm tempted to just merge this with minimal review and deal with whatever fallout happens later. @ppy/team-client objections?

peppy commented:
Yeah I can agree with that. This *seems* to be doing the correct thing. Given how localised storyboard code is I'm inclined to get it in. Worst case we revert or rewrite the whole thing.

bdach reviewed (approved):
yolo
